src/cacheHttpClient.ts

```diff
@@ -83,46 +97,127 @@ async function pipeResponseToStream(
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    const downloadResponse = await httpClient.get(cacheEntry.archiveLocation!);
+    const downloadResponse = await httpClient.get(archiveLocation);
     await pipeResponseToStream(downloadResponse, stream);
 }
 
+// Returns Cache ID
+async function reserveCache(
+    restClient: RestClient,
+    key: string
+): Promise<number> {
+    const reserveCacheRequest: ReserveCacheRequest = {
+        key
+    };
+    const response = await restClient.create<ReserverCacheResponse>(
+        "caches",
+        reserveCacheRequest
+    );
+
+    return response?.result?.cacheId || -1;
+}
+
+function getContentRange(start: number, length: number): string {
+    // Format: `bytes start-end/filesize
+    // start and end are inclusive
+    // filesize can be *
+    // For a 200 byte chunk starting at byte 0:
+    // Content-Range: bytes 0-199/*
+    return `bytes ${start}-${start + length - 1}/*`;
+}
+
+async function uploadChunk(
+    restClient: RestClient,
+    cacheId: number,
+    data: Buffer,
+    offset: number
+): Promise<IRestResponse<void>> {
+    const requestOptions = getRequestOptions();
+    requestOptions.additionalHeaders = {
+        "Content-Type": "application/octet-stream",
+        "Content-Range": getContentRange(offset, data.byteLength)
+    };
+
+    return await restClient.update(
+        cacheId.toString(),
+        data.toString("utf8"),
+        requestOptions
+    );
+}
+
+async function commitCache(
+    restClient: RestClient,
+    cacheId: number,
+    filesize: number
+): Promise<IRestResponse<void>> {
+    const requestOptions = getRequestOptions();
+    const commitCacheRequest: CommitCacheRequest = { size: filesize };
+    return await restClient.create(
+        cacheId.toString(),
+        commitCacheRequest,
+        requestOptions
+    );
+}
+
 export async function saveCache(
     key: string,
     archivePath: string
 ): Promise<void> {
-    const stream = fs.createReadStream(archivePath);
-
-    const cacheUrl = getCacheUrl();
     const token = process.env["ACTIONS_RUNTIME_TOKEN"] || "";
     const bearerCredentialHandler = new BearerCredentialHandler(token);
 
-    const resource = `_apis/artifactcache/cache/${encodeURIComponent(key)}`;
-    const postUrl = cacheUrl + resource;
-
-    const restClient = new RestClient("actions/cache", undefined, [
+    const restClient = new RestClient("actions/cache", getCacheApiUrl(), [
         bearerCredentialHandler
     ]);
 
-    const requestOptions = getRequestOptions();
-    requestOptions.additionalHeaders = {
-        "Content-Type": "application/octet-stream"
-    };
+    // Reserve Cache
+    const cacheId = await reserveCache(restClient, key);
+    if (cacheId < 0) {
+        throw new Error(`Unable to reserve cache.`);
+    }
 
-    const response = await restClient.uploadStream<void>(
-        "POST",
-        postUrl,
-        stream,
-        requestOptions
+    // Upload Chunks
```

Comment:
Part of the reason for this API change was so that we could attempt to reserve the cache before even tarring up the cache. Are we still planning to do that?

Reply:
Yep, I'll move the reserve call to before the tar creation